pyControl4/lock.py

```diff
@@ -0,0 +1,86 @@
+class C4Lock:
+    def __init__(self, C4Director, item_id):
+        """Creates a Control4 Lock object.
+
+        Parameters:
+            `C4Director` - A `pyControl4Ex.director.C4Director` object that
+            corresponds to the Control4 Director that the Lock is connected to.
+
+            `item_id` - The Control4 item ID of the Lock.
+        """
+        self.director = C4Director
+        self.item_id = item_id
+
+    async def is_locked(self):
+        """Returns True if it is currently Locked or False if it is Unlocked
+
+        Notes:
+            Relay State 0 == Locked
+            Relay State 1 == Unlocked
+        """
+        return not bool(
+            await self.director.getItemVariableValue(self.item_id, "RelayState")
+        )
+
+    async def state_verified(self):
+        """Returns True if Lock Relay is functional
+
+        Notes:
+            I think this is just used to verify that the relay is functional,
+            not 100% sure though.
+        """
+        return bool(
+            await self.director.getItemVariableValue(self.item_id, "StateVerified")
+        )
+
+    async def lock(self):
+        """Locks the current lock
+
+         I assume the 'OPEN' command is referring to setting the relay to it's
+         Open state which actually engages the lock
```

Comment:
Is this device represented as a _lock_ in Control4 or is it just a generic _relay_ that happens to be named like a lock? If this code works for all relays then we should probably rename it to `C4Relay` and make all the function names more generic (`open` and `close` instead of `lock` and `unlock`)

Reply:
I am not sure to be honest, I just started working on this as a personal project. I haven't done a ton with C4, but am learning. So I ran `director.getItemInfo()` and this was part of the output `"protocolControl":` `"control4_relaysingle"`, so it would be appropriate to rename it to C4Relay then.

Reply:
Ah I see, it does appear to be a relay device. Could you post the entire output of `director.getItemInfo()` so I can learn a bit more about what this device is? (don't forget to check for and remove sensitive info just in case)

Reply:
I just pushed a commit that changes it to C4Relay, luckily there is no sensitive info for this.

```
[
 {
  "type": 7,
  "typeName": "device",
  "name": "Door Lock",
  "parentId": 305,
  "id": 307,
  "filename": "relaysingle_doorlock_c4.c4i",
  "proxy": "relaysingle_doorlock_c4",
  "proxyOrder": 1,
  "proxyMeta": [
   {
    "name": "relaysingle_doorlock_c4",
    "smallImage": "devices_sm/relaysingle_doorlock_open.gif",
    "smallSource": "composer",
    "largeImage": "devices_lg/relaysingle_doorlock_open.gif",
    "largeSource": "composer"
   }
  ],
  "categories": [
   "locks"
  ],
  "siteId": 2,
  "siteName": "Home",
  "buildingId": 3,
  "buildingName": "House",
  "floorId": 13,
  "floorName": "First Floor",
  "roomId": 305,
  "roomName": "Front",
  "largeImage": "devices_lg/relaysingle_doorlock_open.gif",
  "smallImage": "devices_sm/relaysingle_doorlock_open.gif",
  "protocolId": 307,
  "protocolName": "Door Lock",
  "protocolFilename": "relaysingle_doorlock_c4.c4i",
  "capabilities": {},
  "events": [
   {
    "id": 1,
    "description": "When the NAME is locked"
   },
   {
    "id": 2,
    "description": "When the NAME is unlocked"
   }
  ],
  "commands": {
   "command": [
    {
     "id": 0,
     "description": "Lock the NAME"
    },
    {
     "id": 1,
     "description": "Unlock the NAME"
    },
    {
     "id": 2,
     "description": "Toggle the NAME"
    }
   ]
  },
  "conditionals": {
   "conditional": [
    {
     "id": 0,
     "description": "the NAME is locked"
    },
    {
     "id": 1,
     "description": "the NAME is unlocked"
    }
   ]
  },
  "control": "control4_relaysingle",
  "version": 2,
  "created": "1/20/2005 2:22:54 PM",
  "modified": "11/02/2005 12:30 PM",
  "model": "Door Lock",
  "manufacturer": "Generic",
  "protocolControl": "control4_relaysingle",
  "deviceOrder": 2,
  "allControlBindingsHidden": 0,
  "URIs": {
   "variables": "/api/v1/items/307/variables",
   "bindings": "/api/v1/items/307/bindings",
   "commands": "/api/v1/items/307/commands"
  }
 }
]
```

Reply:
That relay id is 307 on my system, and I have a device named Yale Real Living which is the actual deadbolt lock itself and that id is 306 and here is the `director.getItemInfo()` for the lock. When I locked the deadbolt both device id 307 and 306 both updated saying that it was now locked.

```
[
 {
  "type": 7,
  "typeName": "device",
  "name": "Yale Real Living",
  "parentId": 305,
  "id": 306,
  "filename": "Lock_ZigBee_Yale.c4i",
  "proxy": "Lock_ZigBee_Yale",
  "proxyOrder": 1,
  "proxyMeta": [
   {
    "name": "Lock_ZigBee_Yale",
    "smallImage": "devices_sm/relaysingle_doorlock_open.gif",
    "smallSource": "composer",
    "largeImage": "devices_lg/relaysingle_doorlock_open.gif",
    "largeSource": "composer"
   }
  ],
  "categories": [
   "locks"
  ],
  "siteId": 2,
  "siteName": "Home",
  "buildingId": 3,
  "buildingName": "House",
  "floorId": 13,
  "floorName": "First Floor",
  "roomId": 305,
  "roomName": "Front",
  "largeImage": "devices_lg/relaysingle_doorlock_open.gif",
  "smallImage": "devices_sm/relaysingle_doorlock_open.gif",
  "protocolId": 306,
  "protocolName": "Yale Real Living",
  "protocolFilename": "Lock_ZigBee_Yale.c4i",
  "capabilities": {},
  "events": [
   {
    "id": 1,
    "name": "Fault",
    "description": "The lock is in an unknown or fault state."
   },
   {
    "id": 2,
    "name": "Deadbolt Jammed",
    "description": "The deadbolt motor is jammed."
   },
   {
    "id": 3,
    "name": "Keypad Lock",
    "description": "The lock has been locked by keypad."
   },
   {
    "id": 4,
    "name": "Keypad Unlock",
    "description": "The lock has been unlocked by keypad."
   },
   {
    "id": 5,
    "name": "Manual Lock Key",
    "description": "The lock has been locked manually by key cylinder or inside thumb-turn."
   },
   {
    "id": 6,
    "name": "Manual Lock Touch",
    "description": "The lock has been locked manually by touch function (lock and leave)."
   },
   {
    "id": 7,
    "name": "Manual Unlock",
    "description": "The lock has been unlocked manually by key cylinder or inside thumb-turn."
   },
   {
    "id": 8,
    "name": "RF Operate Lock",
    "description": "The lock has been locked by the RF module."
   },
   {
    "id": 9,
    "name": "RF Operate Unlock",
    "description": "The lock has been unlocked by the RF module."
   },
   {
    "id": 10,
    "name": "Auto Lock Operate Locked",
    "description": "Auto relock cycle complete. Lock is locked."
   },
   {
    "id": 11,
    "name": "Lock Reset To Factory Defaults",
    "description": "The lock was reset manually to factory defaults."
   },
   {
    "id": 12,
    "name": "Master Code Changed",
    "description": "The master code was changed at the keypad."
   },
   {
    "id": 13,
    "name": "Duplicate PIN Code Error",
    "description": "A duplicate PIN code was found at the lock when adding a new user."
   },
   {
    "id": 14,
    "name": "Battery Replacement",
    "description": "The lock was reset because the batteries were replaced."
   },
   {
    "id": 15,
    "name": "Handing Cycle Completed Right",
    "description": "The handing cycle has completed (right)."
   },
   {
    "id": 16,
    "name": "Handing Cycle Completed Left",
    "description": "The handing cycle has completed (left)."
   },
   {
    "id": 17,
    "name": "RF Module Power Cycled",
    "description": "Power to the RF module has been restored."
   },
   {
    "id": 18,
    "name": "Tamper Alarm Keypad Attempts",
    "description": "Keypad attempts exceed the code entry limit."
   },
   {
    "id": 19,
    "name": "Tamper Alarm Front Escutcheon",
    "description": "The front escutcheon has been removed from main."
   },
   {
    "id": 20,
    "name": "Tamper Alarm Low Battery",
    "description": "The battery level is low."
   },
   {
    "id": 21,
    "name": "Tamper Alarm Critical Battery",
    "description": "The battery level is critical."
   },
   {
    "id": 22,
    "name": "Low Battery Alarm",
    "description": "The battery level is too low for the lock to operate."
   },
   {
    "id": 99,
    "name": "Unknown",
    "description": "The lock generated an unknown alarm or error code. Contact technical support."
   }
  ],
  "control": "lua_gen",
  "version": 223,
  "created": "09/14/2010 00:00:00",
  "modified": "6/8/2017 13:19:35",
  "model": "Real Living",
  "manufacturer": "Yale",
  "protocolControl": "lua_gen",
  "deviceOrder": 1,
  "allControlBindingsHidden": 0,
  "isOnline": true,
  "URIs": {
   "variables": "/api/v1/items/306/variables",
   "properties": "/api/v1/items/306/properties",
   "network": "/api/v1/items/306/network",
   "bindings": "/api/v1/items/306/bindings",
   "commands": "/api/v1/items/306/commands"
  }
 }
]

```